### PR TITLE
13 user auth post

### PR DESCRIPTION
**Pull request by benjaminsinaiko**
#13 Created auth/login route

Authentication for User is in the model, similar to the user (register) route.  Uses `UserSchema.statics.authenticate` for authentication and error handling.

Validation gives generic error for bad email or password.  This is hard coded now, may want to look into mongoose error generation.

### CODE REVIEW

File: server/routes/api/auth.js [hunk 1]
```diff
@@ -0,0 +1,25 @@
+const express = require('express');
+const router = express.Router();
+const bcrypt = require('bcrypt');
+
+// User Model
+const User = require('../../models/user');
+
+// POST User Login
+router.post('/login', (req, res) => {
+  User.authenticate(req.body.email, req.body.password)
```
Comment (donthedeveloper) [inline server/routes/api/auth.js]: Although it's not necessarily needed here, it's a good practice to explicitly tell Express that we're done with this route by *returning* this promise.

File: server/routes/api/auth.js [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+    .then(
+      user => {
+        req.session.userId = user._id;
+        res.status(200).json(user);
+      },
+      err => {
+        res.status(500).json(err);
```
Comment (donthedeveloper) [inline server/routes/api/auth.js]: 500 is usually a server problem. Let's return a 400 here. Also, we're almost mimicking that basic structure. We'll want to place the error object in a property of 'error', so what will be returned is:

```
{
    "error": {
      "message": "Incorrect username and password combination.",
      "name": "AuthenticationError"
    }
}

File: server/models/user.js [hunk 3]
```diff
@@ -37,22 +37,38 @@ UserSchema.plugin(uniqueValidator, {
...
-      .catch((err) => next(err))
+      .catch(err => next(err));
   } else {
     next();
   }
 });
 
+UserSchema.statics.authenticate = function(email, password) {
+  return User.findOne({ email })
+    .exec()
```
Comment (donthedeveloper) [inline server/models/user.js]: Is `.exec()` needed here or does `.findOne()` return a promise? I'm genuinely curious.

File: server/routes/api/auth.js (hunk 2 above)
Comment (benjaminsinaiko) [reply server/routes/api/auth.js]: Changed error status to 400 and placed error object in "error" property.

File: server/models/user.js (hunk 3 above)
Comment (benjaminsinaiko) [reply server/models/user.js]: Still not totally clear on `.then()` vs `.exec()`, but it does seem like `.exec()` is needed with mongoose to return an actual promise.  The mongoose docs are a bit confusing, let me know what you think:

>A query also has a .then() function, and thus can be used as a promise.

But then,

>Mongoose queries are not promises. They have a .then() function for co and async/await as a convenience. However, unlike promises, calling a query's .then() can execute the query multiple times. [...] If you need a fully-fledged promise, use the .exec() function.

**Pull request merged**